reference: use distribution reference and remove fork

dmcgowan opened the pull request:
Remove forked reference package. Use normalized named values everywhere and familiar functions to convert back to familiar strings for UX and storage compatibility. The `reference.Named` type is now always normalized in Docker.

WIP: Also requires https://github.com/docker/swarmkit/pull/1884 to keep parsing an non-normalized name from returning an error.

Ping @aaronlehmann @stevvooe

dmcgowan committed:
Update distribution vendor

Signed-off-by: Derek McGowan <derek@mcgstyle.net> (github: dmcgowan)

stevvooe commented:
LGTM

dmcgowan commented:
After this has finished its initial code review, we will merge the swarmkit PR and update the vendor in this PR for merge.

stevvooe commented on cli/command/service/list.go:
```diff
@@ -129,12 +129,12 @@ func printTable(out io.Writer, services []swarm.Service, running, tasksNoShutdow
+			namedTagged, ok := ref.(reference.NamedTagged)
 			if ok {
-				image = namedTagged.Name() + ":" + namedTagged.Tag()
+				image = reference.FamiliarName(namedTagged) + ":" + namedTagged.Tag()
```
Shouldn't this use `WithTag`?

aaronlehmann replied:
It would have to use `reference.FamiliarName(reference.WithTag(reference.TrimNamed(namedTagged), namedTagged.Tag()))`

I think the string manipulation approach is probably be fine, since this is being formatted for display rather than any internal use.

aaronlehmann commented on api/server/router/plugin/plugin_routes.go:
```diff
@@ -46,39 +45,30 @@ func parseHeaders(headers http.Header) (map[string][]string, *types.AuthConfig)
-	// NOTE: Using distribution reference to handle references
-	// containing both a name and digest
-	remoteRef, err := distreference.ParseNamed(remote)
+	var remoteRef reference.Named
```
No need to declare this variable.

dmcgowan committed:
Use distribution reference

Remove forked reference package. Use normalized named values
everywhere and familiar functions to convert back to familiar
strings for UX and storage compatibility.

Enforce that the source repository in the distribution metadata
is always a normalized string, ignore invalid values which are not.
Update distribution tests to use normalized values.

Signed-off-by: Derek McGowan <derek@mcgstyle.net> (github: dmcgowan)

aaronlehmann commented on client/image_push.go:
```diff
@@ -17,30 +17,32 @@ import (
 // It's up to the caller to handle the io.ReadCloser and close it properly.
 func (cli *Client) ImagePush(ctx context.Context, ref string, options types.ImagePushOptions) (io.ReadCloser, error) {
-	distributionRef, err := distreference.ParseNamed(ref)
+	distributionRef, err := reference.ParseNormalizedNamed(ref)
```
Might be good to rename `distributionRef`.

aaronlehmann commented on daemon/cluster/cluster.go:
```diff
@@ -829,50 +827,39 @@ func (c *Cluster) GetServices(options apitypes.ServiceListOptions) ([]types.Serv
-// TODO(nishanttotla): After the packages converge, the function must
-// convert distreference.Named -> distreference.Canonical, and the logic simplified.
+// and returns the image pinned to a digest, such as name@sha256:34234
+// TODO(dmcgowan): return named type instead of string
```
This returns something that's inserted directly into a string field that's part of a gRPC message definition, so I don't think it matters either way.

aaronlehmann commented on daemon/cluster/cluster.go:
```diff
@@ -829,50 +827,39 @@ func (c *Cluster) GetServices(options apitypes.ServiceListOptions) ([]types.Serv
 	}
+	namedRef, ok := ref.(reference.Named)
+	if !ok {
+		return "", errors.New("image reference is an image ID")
```
We're assuming this by the process of elimination. I think it's better to explicitly check that the reference has a digest. We may add other types of non-named references someday.

aaronlehmann commented on daemon/commit.go:
```diff
@@ -228,7 +228,7 @@ func (daemon *Daemon) Commit(name string, c *backend.ContainerCommitConfig) (str
 	imageRef := ""
 	if c.Repo != "" {
-		newTag, err := reference.WithName(c.Repo) // todo: should move this to API layer
+		newTag, err := reference.ParseNormalizedNamed(c.Repo) // todo: should move this to API layer
```
Not sure if it's possible for `c.Repo` to contain `:` or `@` characters, but it might be good to be defensive and return an error if `!IsNameOnly(newTag)`

aaronlehmann commented on daemon/errors.go:
```diff
@@ -14,16 +14,12 @@ func (d *Daemon) imageNotExistToErrcode(err error) error {
 		}
-		tag := reference.DefaultTag
-		ref, err := reference.ParseNamed(dne.RefOrID)
+		ref, err := reference.ParseNormalizedNamed(dne.RefOrID)
```
I think we should change the `RefOrID` to `reference.Reference` instead of reparsing here.

aaronlehmann replied:
Also, it seems like most of this function should be moved to the `Error` method of `ErrImageDoesNotExist`. Then this function would become:

```go
if dne, isDNE := err.(ErrImageDoesNotExist); isDNE {
        return errors.NewRequestNotFoundError(err)
}
return err
```

dmcgowan replied:
There are multiple places I want to change the interfaces to use the reference types, tried to avoid that in this PR to keep down the size.

aaronlehmann replied:
It wasn't possible here before, but is now because references can also contain bare digests. It could be a followup, but I think it's a reasonable thing to do as part of making this change.

dmcgowan replied:
Looked more at where it is used, agreed, updated

aaronlehmann commented on daemon/cluster/executor/container/container.go:
```diff
@@ -132,11 +132,11 @@ func (c *containerConfig) name() string {
 		return raw
 	}
-	return reference.WithDefaultTag(ref).String()
+	return reference.FamiliarString(reference.EnsureTagged(ref))
```
Do we need to check `IsNameOnly` before calling `EnsureTagged`?

aaronlehmann commented on daemon/image_delete.go:
```diff
@@ -232,7 +232,9 @@ func (daemon *Daemon) getContainerUsingImage(imageID image.ID) *container.Contai
-	ref = reference.WithDefaultTag(ref)
+	if reference.IsNameOnly(ref) {
+		ref = reference.EnsureTagged(ref)
+	}
```
Sort of feel like this is a step backwards. Could we add a utility function `EnsureTaggedOrDigested` that does this?

dmcgowan replied:
This is kind of an annoying one, don't really want to make helper functions for the different criteria. Even though at no point do we use ensuretagged when it could be adding to a reference which already has a digest. This is more lines of code, but I think it reads much nicer since it is more obvious how and when the reference is being transformed, rather than having a helper function which reads well enough to understand what it is doing and for the appropriate combinations.

aaronlehmann commented on daemon/image_tag.go:
```diff
@@ -13,12 +13,12 @@ func (daemon *Daemon) TagImage(imageName, repository, tag string) error {
 	}
 
-	newTag, err := reference.WithName(repository)
+	newTag, err := reference.ParseNormalizedNamed(repository)
```
I feel like this function should take a reference instead of assembling it from a string/tag.

Either way, if we switch from `WithName` to `ParseNormalizedNamed`, we should ensure that this is just a name after parsing. I think having a tag or digest inside `repository` should be an error condition, instead of silently being dropped by `TrimNamed` below.

aaronlehmann commented on daemon/import.go:
```diff
@@ -35,7 +35,7 @@ func (daemon *Daemon) ImportImage(src string, repository, tag string, msg string
 	if repository != "" {
 		var err error
-		newRef, err = reference.ParseNamed(repository)
+		newRef, err = reference.ParseNormalizedNamed(repository)
```
TODO to change this to `WithName`?

dmcgowan replied:
Will also add an error here like the previous case, it should change to take in `reference.Named`

aaronlehmann commented on distribution/pull.go:
```diff
@@ -147,7 +147,7 @@ func Pull(ctx context.Context, ref reference.Named, imagePullConfig *ImagePullCo
 		}
 
-		imagePullConfig.ImageEventLogger(ref.String(), repoInfo.Name(), "pull")
+		imagePullConfig.ImageEventLogger(reference.FamiliarString(ref), reference.FamiliarName(repoInfo.Name), "pull")
```
Below this there's `lastErr = fmt.Errorf("no endpoints found for %s", ref.String())` - should this also be converted to `FamiliarString` to preserve the old behavior?

dmcgowan replied:
Yes it should, though not sure if I intentionally left it it out since it mentioned endpoints. Either way this change should not change such behavior.

aaronlehmann commented on distribution/push_v2.go:
```diff
@@ -670,10 +668,9 @@ func numOfMatchingPathComponents(pth string, matchComponents []string) int {
 func getPathComponents(path string) []string {
-	// make sure to add docker.io/ prefix to the path
-	named, err := reference.ParseNamed(path)
+	named, err := reference.ParseNormalizedNamed(path)
```
I don't understand why we're reparsing here. The function is passed `repoInfo.Name()` Couldn't it just take `repoInfo` as an argument and return `strings.Split(repoInfo.Name(), "/")`?

dmcgowan replied:
This is also called by `numOfMatchingPathComponents`, but this could probably be refactored to be more efficient, possibly just guarantee that the normalized string is passed in.

dmcgowan replied:
Pushed changes, haven't resolved this one yet, not quite sure what the best route is. There may be something incorrect about `SourceRepository` right now though, I am not sure it is always normalized.  If it is always normalized I can just make this split and nothing else.

aaronlehmann replied:
It looks like `SourceRepository` is always normalized. The entries are created here:

```go
                        // Cache mapping from this layer's DiffID to the blobsum
                        if err := pd.v2MetadataService.TagAndAdd(diffID, pd.hmacKey, metadata.V2Metadata{
                                Digest:           err.Descriptor.Digest,
                                SourceRepository: pd.repoInfo.FullName(),
                        }); err != nil {
                                return distribution.Descriptor{}, xfer.DoNotRetry{Err: err}
                        }
```

aaronlehmann commented on image/tarexport/save.go:
```diff
@@ -51,16 +51,12 @@ func (l *tarexporter) parseNames(names []string) (map[image.ID]*imageDescriptor,
-					return
-				}
+			tagged, ok := ref.(reference.NamedTagged)
+			if !ok {
```
I don't think the type assertion is necessary.

dmcgowan replied:
the `*imageDescriptor` is expecting a `reference.NamedTagged`

aaronlehmann replied:
I meant you can just call `EnsureTagged` without checking if there's already a tag, but we just changed that function, so you can probably ignore this.

aaronlehmann commented on image/tarexport/save.go:
```diff
@@ -73,28 +69,36 @@ func (l *tarexporter) parseNames(names []string) (map[image.ID]*imageDescriptor,
+				}
+				addAssoc(id, nil)
+				continue
+			} else {
```
Drop the `else` and unindent the `return`.

aaronlehmann commented on image/tarexport/save.go:
```diff
@@ -73,28 +69,36 @@ func (l *tarexporter) parseNames(names []string) (map[image.ID]*imageDescriptor,
 		}
-		if ref.Name() == string(digest.Canonical) {
+
+		if reference.FamiliarName(namedRef) == string(digest.Canonical) {
```
Is this possible? Wouldn't `ParseAnyReference` return a `digestReference` in this case rather than a named reference?

dmcgowan replied:
Only input I could think of is `docker.io/library/sha256:tag`, but not in this case it is doing that right thing... I am going to remove this block

dmcgowan replied:
ooo, you know what, I think this picks up the case where you provide a truncated digest, such as "sha256:abcd", would search to see if there was a corresponding image with that ID.

dmcgowan replied:
This should probably be updated in the future to use `https://godoc.org/github.com/docker/distribution/reference#ParseAnyReferenceWithSet` instead

aaronlehmann replied:
Oh. I forgot we support this. It seems a bit strange to combine a `sha256:` prefix with a truncated digest, but we probably had a good reason to make it work.

> This should probably be updated in the future to use https://godoc.org/github.com/docker/distribution/reference#ParseAnyReferenceWithSet instead

Yes please!

aaronlehmann commented on migrate/v1/migratev1.go:
```diff
@@ -322,7 +323,7 @@ func migrateRefs(root, driverName string, rs refAdder, mappings map[string]image
 		for tag, id := range repo {
 			if strongID, exists := mappings[id]; exists {
-				ref, err := reference.WithName(name)
+				ref, err := reference.ParseNormalizedNamed(name)
```
Should reject references that aren't bare names.

aaronlehmann commented on migrate/v1/migratev1_test.go:
```diff
@@ -40,9 +40,9 @@ func TestMigrateRefs(t *testing.T) {
-		"registry:2": "sha256:2c26b46b68ffc68ff99b453c1d30413413422d706483bfa0f98a5e886266e7ae",
+		"docker.io/library/busybox:latest":                                                                  "sha256:fcde2b2edba56bf408601fb721fe9b5c338d10ee429ea04fae5511b68fbf8fb9",
+		"docker.io/library/busybox@sha256:16a2a52884c2a9481ed267c2d46483eac7693b813a63132368ab098a71303f8a": "sha256:fcde2b2edba56bf408601fb721fe9b5c338d10ee429ea04fae5511b68fbf8fb9",
+		"docker.io/library/registry:2":                                                                      "sha256:2c26b46b68ffc68ff99b453c1d30413413422d706483bfa0f98a5e886266e7ae",
```
What causes the change in behavior?

dmcgowan replied:
Now using normalized strings for maps in memory and this function is doing a deep equal, not comparing against anything that get returned or saved.

dmcgowan replied:
More precisely, it is caused by `*mockTagAddr`, just uses strings.  I could update to use familiar string but I feel that would set a bad precedent for how `Named` values should be indexed by string.

aaronlehmann commented on reference/store.go:
```diff
@@ -44,14 +45,14 @@ type store struct {
 type repository map[string]digest.Digest
 
-type lexicalRefs []Named
+type lexicalRefs []reference.Named
```
`lexicalRefs` and `lexicalAssociations` should be changed to use `FamiliarString` in the `Less` methods, no?

dmcgowan replied:
Maybe can update that later, will keep it the same for now since I am not sure if the strings are getting sorted before presentation.

aaronlehmann commented on daemon/image.go:
```diff
@@ -11,11 +11,11 @@ import (
 
 func (e ErrImageDoesNotExist) Error() string {
-	return fmt.Sprintf("no such id: %s", e.RefOrID)
+	return fmt.Sprintf("no such image: %s", reference.FamiliarString(e.ref))
```
Maybe this should call `TagNameOnly` to preserve the existing error message.

aaronlehmann commented on reference/store.go:
```diff
@@ -54,15 +54,19 @@ type repository map[string]digest.Digest
+func (a lexicalAssociations) Len() int      { return len(a) }
+func (a lexicalAssociations) Swap(i, j int) { a[i], a[j] = a[j], a[i] }
+func (a lexicalAssociations) Less(i, j int) bool {
+	return reference.FamiliarString(a[i].Ref) < reference.FamiliarString(a[j].Ref)
```
Actually, now that I think about it, this is terrible, because every sort comparison has to allocate new strings.

I believe the reason this sorting was added was so that unit tests could more easily check the results of `References` and `ReferencesByName`. Maybe we should change this back to sorting on the normalized name, even though this is a slight change in behavior.

aaronlehmann commented:
@dmcgowan: Tests are failing.

Would be great if you can unsquash (maybe using `git reflog` to dig up original commit hashes if necessary) - it would make the next review round way easier.

aaronlehmann commented:
@dmcgowan: There are swarm test failures.

dmcgowan commented:
@aaronlehmann: failures appear to be related to https://github.com/docker/swarmkit/pull/1884

aaronlehmann commented:
I assume you mean docker/swarmkit#1884 is now necessary. It has been merged to swarmkit master. Please revendor swarmkit.

dmcgowan committed:
Update swarmkit vendor

Signed-off-by: Derek McGowan <derek@mcgstyle.net> (github: dmcgowan)

aaronlehmann commented:
Another test failure :(

dmcgowan commented:
Removed wip tag, the swarmkit update is in, we may proceed when green

aaronlehmann commented:
LGTM

vdemeester reviewed (approved):
That's cool !! 👼
LGTM 🦁
